common.go

```diff
@@ -770,11 +766,20 @@ func (c *Config) time() time.Time {
 
 func (c *Config) cipherSuites(version uint16) []uint16 {
 	if version >= VersionTLS13 {
-		s := c.TLS13CipherSuites
-		if s == nil {
-			s = defaultTLS13CipherSuites()
+	NextCipherSuite:
+		for _, id := range c.CipherSuites {
+			for _, s := range cipherSuites {
+				if s.id != id {
+					continue
+				}
+				if s.flags&suiteTLS13 == 0 {
+					continue NextCipherSuite
+				}
+				return c.CipherSuites
+			}
 		}
-		return s
+		// TODO: this breaks client cipher advertisement.
+		return defaultCipherSuites()
```

Comment:
This is undesirable I think, it ignores the requested CS. I think it is better to reduce maxVersion to 1.2 if no supported TLS 1.3 suite is available. When minVersion is 1.3, then it should become an error. What do you think?

Reply:
@Lekensteyn This is rough and I only put this in here for the sake of demonstrating the proposed feature. If this was accepted, this would *definitely* be fixed.

(For context: "Note: This currently breaks client cipher suite advertisement if version is TLS 1.3, but I believe it was already broken and this can be fixed later if this proposal is accepted.")

That's not a bad idea, but it does go against the express wishes of @FiloSottile. ("I opted for adding a separate TLS13CipherSuites field to the Config because library users that did not set Config.MaxVersion are supposed to get TLS 1.3 support automatically, like it has been for HTTP/2, but having set CipherSuites would effectively disable it.")

To elaborate on why I think this is already broken, `(*Conn).clientHandshake` needs to send both the TLS 1.2 and bellow **and** the TLS 1.3 cipher suites in the client hello. Currently it exclusively sends TLS 1.3 cipher suites when configured for TLS 1.3 (if I'm reading the logic correctly). This is perfectly fine for `(*serverHandshakeState).readClientHello` where only the current version cipher suites are needed.

The fix that seems to make most sense to me is something like this (again rough):
```
NextCipherSuite:
	for _, id := range c.CipherSuites {
		for _, s := range cipherSuites {
			...
		}
	}
	s := c.CipherSuites
	if s == nil {
		return defaultCipherSuites()
	}
	return append(s, defaultTLS13CipherSuites()...)
```

With the above fix, server cipher selection will still work fine (with slight memory overhead, that could be avoided), and the client will offer the configured TLS 1.2 cipher suites followed by the default TLS 1.3 cipher suites.

An alternative would be to only automatically support TLS 1.3 for the server and **not** the client in this case.

Reply:
> Currently [the client] exclusively sends TLS 1.3 cipher suites when configured for TLS 1.3

Indeed broken (in common.go), but TLS 1.3 client support is still work in progress, so the issue has not surfaced yet.

If the concern is enabling TLS 1.3 by default when the user has overridden CipherSuites, what about this adjusted logic (in common.go):

```go
func (c *Config) cipherSuites(version uint16) []uint16 {
	s := c.CipherSuites
	if s == nil {
		s = defaultCipherSuites()
	}
	if c.MaxVersion == 0 || (version >= VersionTLS13 && c.CipherSuites == nil) {
		s = append(s, defaultTLS13CipherSuites())
	}
	return s
}
```
that does not require inspecting the given cipher suites list. Of course if the user sets CipherSuites without any TLS 1.3 ciphers and MaxVersion 1.3, then the handshake will fail. But likewise, if the user sets CipherSuites with only TLS 1.3 ciphers, MinVersion 1.2 would not be that useful.

Reply:
@Lekensteyn I think it would be best to change the API slightly, `(*Config).cipherSuites` is unexported so there are no compatibility problems. That would allow this to work perfectly without requiring any copying from the `append` call.

I honestly really like the idea that expressly setting `c.MaxVersion` to 1.3 and not-including TLS 1.3 ciphers should result in the handshake-failing.